fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,639 @@
+    if current_epoch not in [block_epoch, block_epoch + 1]:
+        return store.finalized_checkpoint.root
+
+    if is_confirmed(store, block_root):
+        return block_root
+    else:
+        return find_confirmed_block(store, block.parent_root)
+
+```
+
+### `get_safe_beacon_block_root``
+
+```python
+def get_safe_beacon_block_root(store: Store) -> Root:
+    head_root = get_head(store)
+
+    return find_confirmed_block(store, head_root)
+```
+
+### `get_safe_execution_payload_hash`
+
+```python
+def get_safe_execution_payload_hash(store: Store) -> Hash32:
+    safe_block_root = get_safe_beacon_block_root(store)
+    safe_block = store.blocks[safe_block_root]
+
+    # Return Hash32() if no payload is yet justified
+    if compute_epoch_at_slot(safe_block.slot) >= BELLATRIX_FORK_EPOCH:
+        return safe_block.body.execution_payload.block_hash
+    else:
+        return Hash32()
+```
+
+*Note*: This helper uses beacon block container extended in [Bellatrix](../specs/bellatrix/beacon-chain.md).
+
+## Confirmation Score
+
+The confirmation score for a block is the maximum adversarial percentage weight that a confirmed block can tolerate.
+This section specifies the algorithm to calculate the confirmation score of a given block.
+This under the assumption that the adversary is willing to get as much stake as possible slashed to
+prevent a block from being confirmed.
+
+### Helper Functions
+
+#### `compute_one_confirmation_score`
+
+```python
+def compute_one_confirmation_score(store: Store, block_root: Root) -> int:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    parent_block = store.blocks[block.parent_root]
+    support = int(get_weight(store, block_root))
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    maximum_support = int(
+        get_committee_weight_between_slots(justified_state, Slot(parent_block.slot + 1), current_slot)
+    )
+    proposer_score = int(get_proposer_score(store))
+
+    # Return the max possible one_confirmation_score such that:
+    # support / maximum_support > \
+    #     0.5 * (1 + proposer_score / maximum_support) + one_confirmation_score / 100
+
+    return max((100 * support - 50 * proposer_score - 1) // maximum_support - 50, UNCONFIRMED_SCORE)
+```
+
+#### `compute_lmd_confirmation_score`
+
+```python
+def compute_lmd_confirmation_score(store: Store, block_root: Root) -> int:
+    if block_root == store.finalized_checkpoint.root:
+        return MAX_CONFIRMATION_SCORE
+    else:
+        block = store.blocks[block_root]
+        finalized_block = store.blocks[store.finalized_checkpoint.root]
+        if block.slot <= finalized_block.slot:
+            # This block is not in the finalized chain.
+            return UNCONFIRMED_SCORE
+        else:
+            # Check one_confirmed score for this block and LMD_confirmed score for the preceding chain.
```

Comment:
Can the recursion break when the slot span covers the full validator set, likewise, the confirmation rule does?

Reply:
Yep. Just implemented.